### DISCUSSION TITLE
뒤로가기 버튼 기능을 구현하기 위해 useNavigate를 사용한 이유

### DISCUSSION THREAD

**Discussion by July249** [beyondDevelops/AnimalTalk#181] [General]
## 뒤로가기 버튼 기능 구현을 위해 useNavigate를 사용한 이유

다른 url로의 이동으로 Link 컴포넌트를 사용할 수 있으나, 이는 특정 url 경로로의 이동을 위한 react-router-dom에서 제공하는 API 입니다. 또한, 이는 브라우저에서 a 요소와 동일하게 취급되기 때문에 시멘틱한 html 작성을 위해 지양되는 경우가 발생합니다.

이러한 Link 컴포넌트의 특징에 유의하여 사용하여야 하며, 동시에 우리가 구현하고자 하였던 Header의 뒤로가기 기능을 제대로 구현하기 위해서는 useNavigate를 사용하여야 했습니다.

## useNavigate 사용방법 (버튼 클릭 이벤트로 동작시키는 경우)
```
import { useNavigate } from 'react-router-dom';

const navigate = useNavigate();

// ...

<button onClick={() => navigate('/login')}>로그인 페이지로 이동</button>
```

## navigate 인자 ( navigate(to, { state={ "any" }, replace="boolean" }) )
- to
   - navigate의 첫번째 인자인 to는 생략이 가능하며 이동하고자 하는 url을 입력하면 됩니다. (root로부터의 상대경로로 지정합니다.)
   - Link의 to 인자와 동일합니다.

- state
  - to로 입력한 경로로 이동하면서 state에 전달하고자 하는 파라미터를 입력하여 이동한 경로에서도 state에 저장된 데이터를 전달받을 수 있습니다. 

- replace
  - boolean 타입으로 값을 정할 수 있으며 기본값은 false 이며, true로 지정시 해당 url로 이동시 히스토리를 남기지 않게 됩니다.
  - 이러한 특성을 이용하면 공식문서의 예시와 같이 로그인하지 않은 사용자가 장바구니 페이지로 이동하고자 한다면 로그인 페이지로 이동시켜주고, 이동한 로그인 페이지에서는 뒤로가기 버튼을 눌러도 장바구니 페이지로 이동되지 않게 됩니다.  

- navigate(-1)을 이용하면, 사용자가 방문하였던  바로 직전 url로 이동하게 됩니다. 이는 뒤로가기 버튼의 기능과 동일합니다.

공식문서 (https://reactrouter.com/en/6.6.1/hooks/use-navigate)

Note: react-router-dom의 6 버전에서는 더이상 useHistory를 지원하지 않습니다. 대신 useNavigate hook을 제공합니다.

**Comment by IntHyun** [comment]
고생하셨습니다!
